text/0232-maturity-of-otel.md

```diff
@@ -0,0 +1,153 @@
+
+#### Development
+
+Not all pieces of the component are in place yet, and it might not be available for users yet. Bugs and performance issues are expected to be reported. User feedback is desired, especially regarding the user experience (configuration options, component observability, technical implementation details, ...). Configuration options might break often depending on how things evolve. The component SHOULD NOT be used in production.
+
+#### Alpha
+
+This is the default level: any components with no explicit maturity level should be assumed to be "Alpha". The component is ready to be used for limited non-critical production workloads, and the authors of this component welcome user feedback. Bugs and performance problems are encouraged to be reported, but component owners might not work on them immediately. The configuration options might often change without backward compatibility guarantees.
+
+#### Beta
+
+Same as Alpha, but the interfaces (API, configuration, generated telemetry) are treated as stable whenever possible. While there might be breaking changes between releases, component owners should try to minimize them. A component at this stage is expected to have had exposure to non-critical production workloads already during its Alpha phase, making it suitable for broader usage.
+
+#### Stable
+
+The component is ready for general availability. Bugs and performance problems should be reported, and there's an expectation that the component owners will work on them. Breaking changes, including configuration options and the component's output, are not expected to happen without prior notice unless under special circumstances.
+
+#### Deprecated
+
+Development of this component is halted, and no further support will be provided. No new versions are planned, and the component might be removed from its included distributions. Note that new issues will likely not be worked on except for critical security issues. Components that are included in distributions are expected to exist for at least two minor releases or six months, whichever happens later. They also MUST communicate in which version they will be removed.
+
+#### Unmaintained
+
+A component identified as unmaintained does not have an active code owner. Such components may have never been assigned a code owner, or a previously active code owner has not responded to requests for feedback within 6 weeks of being contacted. Issues and pull requests for unmaintained components SHOULD be labeled as such. After 6 months of being unmaintained, these components MAY be deprecated. Unmaintained components are actively seeking contributors to become code owners.
+
+### Specification requirements
+
+The specification should declare all requirements for a specified component to be declared stable. When implementations are still missing features required by the specification, they can be at most Alpha status.
+
+### Tier-1 SIGs
```

Comment:
What's the criteria for a Tier-1 SiG vs. a Tier-1 Components?

For component, I think I can suggest a litmus test. For a SiG (group of people) I'm not sure I have an answer.


For example, Looking at this list, I'd argue we should target the following:

- Specification (all components marked stable)
- Semantic Conventions (similarly to the above)
- OTLP/Protocol (components marked stable)
- Collector-core distribution
- Collector-contrib distribution
- SDK distributions (Java, .NET, Go, JavaScript, Python)
- Instrumentation components deemed critical by Language SiGs (e.g. Java auto-agent)

Effectively, we ask ourselves:

- What are the bare minimum components needed for successful opentelemetry adoption?
- What pieces are needed for ongoing success of OpenTelemetry users?


Things like OpAMP we can exclude for now as the maturity and need rises.

Reply:
> For example, Looking at this list, I'd argue we should target the following:

From that list, I would argue that the contrib distribution should _not_ be part of it, as it includes components that are unlikely to be supported. This is why I think SIGs should list what they believe to be the right tier-1 deliverables. 

What I like about fixing the list of components here is that we know right now the list of components we'll aim for stabilizing before anything else.

Reply:
This is why I think the overall list should be COMPONENTS, not SiGs.   I think it's possible if we classify SiGs we'll get some unfairness/bias against smaller SiGs that *are* ready to participate in stability of OTEL.

I'd rather see an approach where we do something like the following:

1. For each "component" in OTEL, it can only declare itself stable if it implements a *stable* release of all *stable* sections of the specification.
2. For utlity components like the collector, we require a strict separate of stable from unstable.   For example, I actually don't think we can consider the metrics portion of the Collector "done" until Prometheus support is stable.  While there is no specification to control the collector, I think we need some form of document denoting what we consider *key* or *minimum viable* functionality that collector must have to be considered stable.   Similarly, we need such a definition from the OTEL Operator, e.g.
3. We work w/ SiGs / Maintainers to understand which are able to comply with stability.   Areas of OTEL that are not participating will need to be clearly denoted, possibly via some form of convention.   E.g. if opentelemetry-rust *is* ready to participate in stability, it can do so simply by having stable components released.  If e.g. Haskell (https://github.com/ethercrow/opentelemetry-haskell) wants to be part of OTEL, it would be held to the same standards and can participate in the "stable" portion of otel.  _Note:  I think it's ok for us to put requirements on active community size._

Right now though, with this proposal, I'm not sure I could figure out what the criteria or value-function for deciding "tier-1 SiGs" would be.

Reply:
> I think we need some form of document denoting what we consider key or minimum viable functionality that collector must have to be considered stable.

Who's "we" here? I take it that you mean the TC, while this OTEP suggests it's up for the Collector maintainers to decide. 

> it can do so simply by having stable components released

One problem I saw with that is that Rust (sorry) could have a stable release there, but if one specific maintainer leaves the project, it leaves the project in a state where we cannot call stable anymore. That's why I think the willingness of a language/SIG has to count, but the ultimate call is on the TC.

> what the criteria or value-function for deciding "tier-1 SiGs" would be

The "Promoting SIGs and components" outlines the process but purposely not the criteria. It's left to the TC members to decide based on available information. If you think a set of criteria should be define beforehand, we could certainly come up with some, although I would really prefer not to have this list until we have gone through this a couple of times.

This all said, I'm leaning more towards listing components here in this OTEP and not SIGs. I think classifying SIGs as tier-1 has caused more harm than good.